libraries/stdlib/src/kotlin/io/Console.kt

```diff
@@ -171,11 +170,36 @@ private val stdin: BufferedReader by lazy { BufferedReader(InputStreamReader(obj
     public override fun read(b: ByteArray, off: Int, len: Int): Int {
         return System.`in`.read(b, off, len)
     }
-}))}
+}}
 
 /**
  * Reads a line of input from the standard input stream.
  *
  * @return the line read or `null` if the input stream is redirected to a file and the end of file has been reached.
  */
-public fun readLine(): String? = stdin.readLine()
+public fun readLine(): String? {
+    // Writing line into ByteArray because InputStreamReader uses buffer
+    val bytes = ByteArrayOutputStream().use { outputStream ->
+        val inputStream = stdin
+        val endOfStream = -1
+        val carriageReturn = '\r'.toInt()
+        val lineFeed = '\n'.toInt()
+
+        var previous = inputStream.read()
+        if (previous == endOfStream) return null
+        if (previous == lineFeed) return ""
+        var current = inputStream.read()
+
+        while (current != endOfStream && current != lineFeed) {
+            outputStream.write(previous)
+            previous = current
+            current = inputStream.read()
+        }
+
+        if (!(previous == carriageReturn && current == lineFeed)) outputStream.write(previous)
+
+        outputStream.toByteArray()
+    }
+
+    return ByteArrayInputStream(bytes).reader().use { it.readText() }
```

Comment:
We could just convert these bytes to string with the `ByteArray.toString(encoding)` function.

By the way, I wonder which encoding is used to encode characters in `System.in`?

Reply:
Unfortunately it is not going to work with UTF-8 as it could contain multi-byte characters with 0x0d or 0x0a inside. For example http://www.fileformat.info/info/unicode/char/020d/index.htm or even http://www.fileformat.info/info/unicode/char/040d/index.htm

Reply:
Perhaps we could use [this solution](https://stackoverflow.com/questions/17227331/decode-bytes-to-chars-one-at-a-time) to decode chars one at a time in the given encoding, though I'm concerned with how much garbage it would produce.

Reply:
@ilya-g Thanks!
Default Charset in InputStreamReader is platform dependent but `InputStream.reader()` in Kotlin uses `Charsets.UTF_8` as default. For using platform dependent charset much more complicated algorithm is needed. I'm not sure it's worth it so I'll use UTF-8 for now.
I don't think it's necessary to read chars right away. We always can translate needed chars to bytes and work with them. Especially considering that we always have to check `read()` output for -1.
@cy6erGn0m If I understand UTF-8 encoding correctly no other char can contain `\r` or `\n` chars. Their binary representations are `0000 1101` and `0000 1010` respectively and any bytes in multi byte char should start with at least one `1`.
The site with examples is down so I can't check them for now.

Reply:
>  I'm not sure it's worth it so I'll use UTF-8 for now.

I don't think it's a matter of choice. Making `readLine` only work with UTF-8 default encoding will lead to simply incorrect chars being read from Windows console for example.

Reply:
@Meztihn to work properly with any encoding you could try to adopt something like this: https://gist.github.com/cy6erGn0m/464453a2059c11f9b31b949aacc3b8ce (feel free to copy it if needed)

Also it is important to write tests: you can use old-school PipedInputStream/PipedOutputStream to verify corner-cases